Added a transformer that injects byte sequences

blanu opened the pull request:
This transformer has a list of byte sequences to add and another list to remove.
Each sequence also has an index into the packet stream, an offset into the packet, and a target packet length.
So for instance, packet #4 could have a byte sequence injected at the 17th byte of a packet of length 118.
An injectable packet is constructed by placing the byte sequence at the specified offset in the packet and then filling in the rest with random bytes in order to generate a packet of the target length.
As the index of the packet is also specified, injected packets will always be sent in a specified order, interleaves with real data packets.

On the receiving end, packets may have been reordered. Therefore, the receiver determines which packets to remove by matching byte sequences to offsets and ignoring the index. Since there are no constraints on the injected byte sequences, false positives are possible. For instance, a packet could be injected where the first byte is 0. This is a common value for the first byte, so based on sequence matching a non-injected packet may be dropped. In this case, SCTP retransmission should be sufficient to recover. False positives are unlikely to occur in practice using real settings for the byte sequences to inject.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-lib/279)

<!-- Reviewable:end -->

Commit:
Added a transformer that injects byte sequences

trevj commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn8FmKHBKQpnQbkX4l)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L9)):</sup>
Is this the object expected, in serialised form, by `ByteSequenceShaper.configure`?

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynCAoDPItxC4wIBMqP)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L60)):</sup>
This isn't an extremely helpful comment and this variable is modified by >1 function. What exactly does it represent?

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 67 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn8nTx0wx8LjA9r0Fz)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L67)):</sup>
I just remembered it's good practice to have unimplemented methods raise an error (fail fast, and all that), e.g.:
```typescript
public setKey = (key:ArrayBuffer) : void => {
  throw new Error('setKey unimplemented');
}
```

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 76 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn9bzZGeS8rLT57mZD)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L76)):</sup>
Confused by this; does this expect a serialised `SequenceConfig`?

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 87 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynA-6H2cjgsTJ0lX1o)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L87)):</sup>
Logging and throwing is generally a code smell: just throw the error and let the caller decide how/if to handle it.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 92 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn9l9QjHQaslpf3iu0)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L92)):</sup>
Don't swallow errors; remove this catch, allowing the caller decide how to handle an error.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynBL0AN1cLjYHOMedi)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L100)):</sup>
style: this brace can fit on the previous line

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 113 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynBb1K7q4kOKoxUk2K)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L113)):</sup>
You're calling `this.outputIndex_++` in both the if and else blocks. Suggest you move it to just before this if/else block.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 145 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynB2j9TdiQ5Ez-h0OG)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L145)):</sup>
Prefer static functions when possible, to reduce state. This looks like it can be static, as well as the functions it calls.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279)
<!-- Sent from Reviewable.io -->

Commit:
Added better documentation of the configuration interface types.

Commit:
Added additional documentation for the ByteSequenceShaper outputIndex_ variable.

Commit:
Throw "not implemented" exception when setKey is called on ByteSequenceShaper

Commit:
Check for the correct keys to see if the configuration json is (obviously) not a SequenceConfig

Commit:
Removed error catching in configure()

Commit:
Moved curly brace to end of line instead of on new line

Commit:
Added some whitespace around operators

Commit:
Refactored packet pushing and outputIndex_ management into a new function outputAndIncrement_

Commit:
Use static for SequenceConfig deserializer functions.

blanu commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn8FmKHBKQpnQbkX4l)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L9)):</sup>
Yes, I have added additional comments to this effect.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynCAoDPItxC4wIBMqP)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L60)):</sup>
I have added additional comments to address this.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 67 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn8nTx0wx8LjA9r0Fz)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L67)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 76 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn9bzZGeS8rLT57mZD)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L76)):</sup>
Okay I changed this.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 87 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynA-6H2cjgsTJ0lX1o)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L87)):</sup>
Okay I changed this.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 92 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jyn9l9QjHQaslpf3iu0)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L92)):</sup>
Okay I changed this.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynBL0AN1cLjYHOMedi)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L100)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 113 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynBb1K7q4kOKoxUk2K)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L113)):</sup>
Okay I have rewritten this.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 145 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynB2j9TdiQ5Ez-h0OG)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/accabe1302bf93646f7447ade27ffa48eddf555e/src/fancy-transformers/byteSequenceShaper.ts#L145)):</sup>
Okay I have made this change.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279)
<!-- Sent from Reviewable.io -->

trevj commented:
Just a couple of things, then :+1:.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 37 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jynxltbbqmqw278ApIe)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L37)):</sup>
I think you had this right the first time, in that it doesn't need to be exported.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 152 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynxqqVsL9PSFdTGSHZ)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L152)):</sup>
spaces: `i < config.addSequences`

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 165 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jynxwe4SQeMAkUq3GUS)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L165)):</sup>
It's more grokkable as an object if you put each field on its own line, i.e.
```typescript
return {
  index: model.index,
  offset: model.offset,
  // ...
};
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279)
<!-- Sent from Reviewable.io -->

Commit:
Added some whitespace around operators

Commit:
Put each field on its own line when making a new SequenceModel

blanu commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 31 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jynxltbbqmqw278ApIe)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L31)):</sup>
It needs to be exported in order for the deserializer methods to be changed from private to static. Static functions are public and so are exported with the class, so all of their argument and return types must be exported as well.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 150 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-JynxqqVsL9PSFdTGSHZ)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L150)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/byteSequenceShaper.ts, line 163 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279#-Jynxwe4SQeMAkUq3GUS)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/b1a768cfc340f627c4ef5c1131e5f73894ff210b/src/fancy-transformers/byteSequenceShaper.ts#L163)):</sup>
Okay I have made this change.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279)
<!-- Sent from Reviewable.io -->

trevj commented:
Still :+1: .

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/279)
<!-- Sent from Reviewable.io -->
